MIXEDARCH-80: Guide on using node affinity to safely schedule pods in heterogeneous clusters. (#5983)
operator-framework/operator-sdk: website/content/en/docs/advanced-topics/multi-arch.md

weight: 80
---

# Supporting Multiple Architectures

Authors may decide to distribute their bundles for various architectures: x86_64, aarch64, ppc64le, s390x, etc, to accommodate the diversity of Kubernetes clusters and reach a larger number of potential users. Each architecture requires however compatible binaries.

## Fundamentals

> The basics of supporting multiple platforms is to ensure that each of your operator images are built for each of the platforms to be supported. From there, the images should be hosted in image registries as manifest lists. Finally, you'll made need to update your distribution configuration to set which platforms are supported. This section explains each of these concepts in turn.

Reviewer: is "platform" the correct term here? maybe "architecture platforms"? or just "architectures" ?

nit: `Finally, you'll need `

That said, the same general advice would apply to supporting workloads on clusters of the same architecture but different operating systems (e.g. `linux/amd64`, `windows/amd64`). In this guide, we assume all workloads will be using the `linux` operating system.

## Fundamentals
The basic principle of supporting multiple architectures is to ensure that each of your operator images is built for each of the architectures to be supported. From there, the images should be hosted in image registries as manifest lists. Finally, you'll need to update your distribution configuration to set which architectures are supported. This section explains each of these concepts in turn.

### Building an Operator for Multiple Architectures

> [Kubebuilder][kubebuilder] explains how you can use `docker buildx` to build images for multiple architectures. Operator SDK leverages KubeBuilder to ensure that builds can be cross-platform from the start.

Reviewer 2: Nit: 
```suggestion
[Kubebuilder][kubebuilder] explains how you can use `docker buildx` to build images for multiple architectures. Operator SDK leverages Kubebuilder to ensure that builds can be cross-platform from the start.
```

> [KubeBuilder explains][kubebuilder] how you can use `docker buildx` to build images for multiple architectures. Operator SDK leverages KubeBuilder to ensure that builds can be cross-architecture from the start.

Reviewer 2: Nit: Kubebuilder shouldn't have a capital B for "builder" in it's name. Remove "explains" from the linking.
```suggestion
[Kubebuilder][kubebuilder] explains how you can use `docker buildx` to build images for multiple architectures. Operator SDK leverages Kubebuilder to ensure that builds can be cross-architecture from the start.
```

Reviewer 2: Also curious about the term "cross-architecture", is that a standard? I'm assuming cross-platform is equivalent to cross-architecture but I've only ever heard the term cross-platform used.

Author: The note at the top of the doc explains why I use the word "architecture" as opposed to "platform" for this guide; however, I don't mind making the change for this specific call-out.

I am thread to thread the needle of balancing when we use platform (os+arch+variant) and when we use architecture. I've erred a little on the side of using architecture for this guide since it is a multi-architecture-specific guide and it's also the aspect that you'll build for most often when talking about workloads for enterprise computing.

 
**Caveat**: the Dockerfile generated by the SDK for the operator explicitly references GOARCH=amd64 for go build. This can be amended to GOARCH=$TARGETARCH. Docker will automatically set the environment variable to the value specified by --platform. With buildah --build-arg will need to be used for the purpose.

**Caveat**: When mirroring registries for disconnected installations (environments without internet connection) all the images referenced by a manifest list need to be copied, including images for architectures that may not be used in the environment.

### Operator Lifecycle Manager

For operators distributed through the [Operator Lifecycle Manager (OLM)][olm]:

* [Bundle images][bundle] are not architecture-specific. They contain only plaintext kubernetes manifests and operator metadata.
* All image references in the ClusterServiceVersion should be manifest lists containing the pointers to the image manifests for the supported architectures.
* Labels for OS and architectures can be set in the CSV. Please refer to the [Operator Lifecycle Management Documentation][olm_multiarch] for details.

## Supporting Clusters with Multi-Architecture Compute Nodes

> The [Fundamentals](#Fundamentals) above aim to guide authors on the key steps to building and distributing operators that can run on multiple architectures. These instructions are sufficient when your cluster's compute nodes share the same supported platform. However, operator authors should also understand the implications of running their operators in a cluster with multi-architecture compute nodes, since it is not always guaranteed that the architectures of the compute nodes will match the architectures supported by the operator.

Reviewer: as long as we make it clear what platform means in this context - i am good with it.

### Safe Scheduling Using Node Affinity

> Node affinity is a mechanism exposed in a Kubernetes pod template that allows a `PodSpec` author to instruct the scheduler to restrict a pod to run only on (or with preference for) nodes that meet specific criteria. In order to ensure that pods are always scheduled to nodes of compatible architecture, it is a best practice for authors to set node affinity requirements to ensure their operators and operands will only schedule to the platforms supported by the images in the pod. If you don't do this, a container scheduled to an incompatible node will immediately crash with an `exec format error`, which will ultimately lead to an `ImagePullBackoff` event as the pod is restarted only to crash again with the same error.

Reviewer: `will only schedule to the platforms`
to
`will only schedule to the nodes with architectures available to the images in the pod`

docker manifest inspect <image>
```

Here's an example of platforms distributed for the Alpine Linux container image. Notice that both linux/amd64 and linux/arm64 are supported.
```shell
$ docker manifest inspect alpine
  "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 528,
         "digest": "sha256:1304f174557314a7ed9eddb4eab12fed12cb0cd9809e4c28f29af86979a3c870",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 528,
         "digest": "sha256:5da989a9a3a08357bc7c00bd46c3ed782e1aeefc833e0049e6834ec1dcad8a42",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v6"
         }
      },
      ...
```

If the above command fails, it may be because the referenced name points to a single image rather than a manifest list. In this case, you can find the architecture by inspecting the image.
```shell

> docker pull <image> && docker inspect <image>

Reviewer: instead of using docker and asking users to pull the image , we could ask them to `skopeo inspect --raw  <image>` first and if there aren't any manifests in the output just do a `skopeo inspect <image>`  to look at the manifest arch.

Author: that sounds cleaner :)
:+1:

### Safe Scheduling Using Node Affinity
Node affinity is a mechanism exposed in a Kubernetes pod template that allows a `PodSpec` author to instruct the scheduler to restrict a pod to run only on (or with a preference for) nodes that meet specific criteria. To ensure that pods are always scheduled to nodes of compatible architecture, it is a best practice for authors to set node affinity requirements to ensure their operators and operands will only schedule to the nodes with architectures available to the images in the pod. If you don't do this, a container scheduled to an incompatible node will immediately crash with an `exec format error`, which will ultimately lead to an `ImagePullBackoff` event as the pod is restarted only to crash again with the same error.

#### Determining the Architectures Supported by an Image
For a given container image, you can check which architectures are supported by listing them by inspecting the manifest. Piping the output to the python `json.tool` module enables pretty-printed JSON output.
```shell

> $ skopeo inspect --raw <image> | python -m 'json.tool'

Reviewer 2: Might also be nice to show how this inspect can be done with `docker inspect ...`

Author: Added this back into this draft after showing how to do it with skopeo. Since the output is basically the same, I just showed the commands.

The list of architecture values should only include the architectures the operator supports. The full list of possible `GOARCH` values is [available here][goarch]. It should be noted that Kubernetes allows a user to specify both `requiredDuringSchedulingIgnoredDuringExecution` and `preferredDuringSchedulingIgnoredDuringExecution`. The syntax for each block is the same. At a minimum, an operator author should set the required terms since they protect against a pod being scheduled on an incompatible node. If the operator in question has much better performance on a subset of the supported architectures, it may also be prudent to set `preferredDuringSchedulingIgnoredDuringExecution` so that the optimal arches are selected first if they are available.

##### Setting Node Affinity in a Kubernetes Manifest

> To update the `PodSpec` and `PodSpecTemlate` objects of an operator and its operands, you will need to scan your operator for instances of these objects. Most of the time, these objects will be defined directly in a Kubernetes manifest `yaml`; however, for dynamically created workloads, these are sometimes embedded in the operator's logic directly.

Reviewer 2: Nit:
```suggestion
To update the `PodSpec` and `PodTemplateSpec` objects of an operator and its operands, you will need to scan your operator for instances of these objects. Most of the time, these objects will be defined directly in a Kubernetes manifest `yaml`; however, for dynamically created workloads, these are sometimes embedded in the operator's logic directly.
```

> The most common directory with `PodSpec` and `PodSpecTemplate` objects that will need to be updated will be in the operator's `config` directory. Additionally, if the operator is configured for OLM, there will likely be a deployment object in the `bundle/manifests/*.clusterserviceversion.yaml`.

Reviewer 2: Nit:
```suggestion
The most common directory with `PodSpec` and `PodTemplateSpec` objects that will need to be updated will be in the operator's `config` directory. Additionally, if the operator is configured for OLM, there will likely be a deployment object in the `bundle/manifests/*.clusterserviceversion.yaml`.
```

            ...
        },
        Containers: []corev1.Container{{
            ...
        }},
    },
```

##### Updating Operator Lifecycle Manager Configurations for Multi-Architecture Compute Nodes
The [Operator Lifecycle Manager (OLM)][olm] is often used to distribute operators via operator catalogs. The operator deployment object in an OLM integrated operator is defined in the `ClusterServiceVersion` yaml. It is important to **remember to set the node affinity block in each of the `spec.template.spec` under `spec.install.deployments`**.

If you're planning to distribute the operator via OLM, you can find more information in the [OLM guide for supporting multiarch][olm_multiarch].

#### Overriding Affinity for an Operator Pod as a Cluster Admin
A cluster admin might have some context that can be used to refine the scheduling requirements for the operator images specified in the deployment spec of the Cluster Service Version. To ensure the fine control of scheduling remains with the cluster admin, OLM provides a mechanism for [overriding the operator pod affinity configuration][olm_affinity_override] as part of the `Subscription` object.

> **Caveat**: No built-in affinity override controls exist that allow cluster admins to override affinity for operands. It is up to the operator author to determine whether this would be appropriate for one or more of their operands what configurations options should be allowed.

Reviewer 2: nit:
```suggestion
**Caveat**: No built-in affinity override controls exist that allow cluster admins to override affinity for operands. It is up to the operator author to determine whether this would be appropriate for one or more of their operands and what configuration options should be allowed.
```
